lib/babe/babe.go

```diff
@@ -336,77 +301,73 @@ func (b *Service) initiate() {
+		epochData,
+		b.constants,
+		b.handleSlot,
+		b.keypair,
+	)
 }
 
-func (b *Service) invokeBlockAuthoring() error {
+func (b *Service) runEngine() error {
 	epoch, err := b.epochState.GetCurrentEpoch()
 	if err != nil {
 		logger.Errorf("failed to get current epoch: %s", err)
 		return err
 	}
 
 	for {
-		err := b.initiateEpoch(epoch)
-		if err != nil {
-			logger.Errorf("failed to initiate epoch %d: %s", epoch, err)
-			return err
-		}
-
-		logger.Debugf("initiated epoch with threshold %s, randomness 0x%x and authorities %v",
-			b.epochData.threshold, b.epochData.randomness[:], b.epochData.authorities)
-
-		epochStartSlot, err := b.waitForEpochStart(epoch)
-		if err != nil {
-			logger.Errorf("failed to wait for epoch %d start: %s", epoch, err)
-			return err
-		}
-
-		// calculate current slot
-		startSlot := getCurrentSlot(b.slotDuration)
-		intoEpoch := startSlot - epochStartSlot
-
-		// if the calculated amount of slots "into the epoch" is greater than the epoch length,
-		// we've been offline for more than an epoch, and need to sync. pause BABE for now, syncer will
-		// resume it when ready
-		if b.epochLength <= intoEpoch && !b.dev {
-			logger.Debugf(
-				"pausing BABE, need to sync since we have %d slots (start slot %d) into the epoch starting at %d",
-				intoEpoch, startSlot, epochStartSlot)
-			return b.Pause()
-		}
-
-		if b.dev {
-			intoEpoch = intoEpoch % b.epochLength
-		}
-
-		logger.Infof("current epoch %d has %d slots", epoch, intoEpoch)
-
-		// get start slot for current epoch
-		nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
-		if err != nil {
-			logger.Errorf("failed to get start slot for next epoch %d: %s", epoch+1, err)
-			return err
-		}
+		err = func() error {
+			ctx, cancel := context.WithCancel(b.ctx)
+			defer cancel()
+			b.epochHandler, err = b.initiateAndGetEpochHandler(ctx, epoch)
+			if err != nil {
+				return err
+			}
 
-		nextEpochStartTime := getSlotStartTime(nextEpochStart, b.slotDuration)
-		epochTimer := time.NewTimer(time.Until(nextEpochStartTime))
-		cleanup := func() {
-			if !epochTimer.Stop() {
-				<-epochTimer.C
+			// get start slot for current epoch
+			nextEpochStart, err := b.epochState.GetStartSlotForEpoch(epoch + 1)
+			if err != nil {
+				logger.Errorf("failed to get start slot for next epoch %d: %s", epoch+1, err)
+				return err
 			}
-		}
 
-		slotDone := make([]<-chan time.Time, b.epochLength-intoEpoch)
-		for i := 0; i < int(b.epochLength-intoEpoch); i++ {
-			slotDone[i] = time.After(b.getSlotDuration() * time.Duration(i))
-		}
+			nextEpochStartTime := getSlotStartTime(nextEpochStart, b.constants.slotDuration)
+			epochTimer := time.NewTimer(time.Until(nextEpochStartTime))
+			cleanup := func() {
+				if !epochTimer.Stop() {
+					<-epochTimer.C
+				}
+			}
 
-		for i := 0; i < int(b.epochLength-intoEpoch); i++ {
-			done := false
+			errCh := make(chan error)
+			go b.epochHandler.run(errCh)
 
 			select {
 			case <-b.ctx.Done():
```

Comment:
I feel like you would want to return for `runEngine` func here, don't that's what we are doing.

(May be we should start passing `ctx` to func as well. That would make the purpose of the `ctx` more obvious)

Reply:
what do you mean? I never want to return from `runEngine` unless the node is shut down

Reply:
I have to initialize the ctx in the function so that the cancel is called when the function returns

Reply:
> I never want to return from runEngine unless the node is shut down

`b.ctx` is context of the service. If `b.ctx` is done, that should mean we don't need the use of service anymore, hence all of its tasks should stop, which does not happen. Is there any other meaning of `b.ctx` here?

But, even if there is any other meaning of `b.ctx`, once `b.ctx` is Done, in the next iteration we end up creating a nested context from a context that is already finished (which is possible, I don't know why). And then without waiting you will do cleanup again because `b.ctx` is already done.
So, it basically becomes an infinite loop of cleanup.

Reply:
ah ok I understand, I've updated the inner function to return an error if `b.ctx` is cancelled, which will then cause the function to return.